src/rust/ide/src/model/project.rs

```diff
@@ -92,78 +92,75 @@ impl ExecutionContextsRegistry {
-pub struct Handle {
-    pub project_name        : Rc<String>,
+#[derive(Debug)]
+pub struct Project {
+    pub name                : Rc<String>,
     pub language_server_rpc : Rc<language_server::Connection>,
-    pub visualization       : Visualization,
     pub language_server_bin : Rc<binary::Connection>,
+    pub visualization       : Visualization,
     pub module_registry     : Rc<model::registry::Registry<ModulePath,model::synchronized::Module>>,
     pub execution_contexts  : Rc<ExecutionContextsRegistry>,
+    pub suggestion_db       : Rc<SuggestionDatabase>,
     pub parser              : Parser,
     pub logger              : Logger,
 }
 
-impl Handle {
-    /// Create a new project controller.
-    pub fn new
+impl Project {
+    /// Create a new project model.
+    pub async fn new
     ( parent              : impl AnyLogger
     , language_server_rpc : Rc<language_server::Connection>
     , language_server_bin : Rc<binary::Connection>
-    , project_name        : impl Str
-    ) -> Self {
+    , name                : impl Str
+    ) -> FallibleResult<Self> {
         let logger = Logger::sub(parent,"Project Controller");
-        info!(logger,"Creating a project controller for project {project_name.as_ref()}");
-        let json_rpc_events        = language_server_rpc.events();
+        info!(logger,"Creating a model of project {name.as_ref()}");
         let binary_protocol_events  = language_server_bin.event_stream();
+        let json_rpc_events         = language_server_rpc.events();
         let embedded_visualizations = default();
         let language_server         = language_server_rpc.clone();
         let visualization           = Visualization::new(language_server,embedded_visualizations);
-        let project_name            = Rc::new(project_name.into());
+        let name                    = Rc::new(name.into());
         let module_registry         = default();
         let execution_contexts      = default();
         let parser                  = Parser::new_or_panic();
+        let language_server         = &*language_server_rpc;
+        let suggestion_db           = SuggestionDatabase::create_synchronized(language_server);
+        let suggestion_db           = Rc::new(suggestion_db.await?);
 
-        let ret = Handle {project_name,module_registry,execution_contexts,parser,
-            language_server_rpc,language_server_bin,logger,visualization};
+        let ret = Project {name,module_registry,execution_contexts,parser,
+            language_server_rpc,language_server_bin,logger,visualization,suggestion_db};
 
         let binary_handler = ret.binary_event_handler();
         crate::executor::global::spawn(binary_protocol_events.for_each(binary_handler));
 
         let json_rpc_handler = ret.json_event_handler();
         crate::executor::global::spawn(json_rpc_events.for_each(json_rpc_handler));
 
-        ret
+        ret.acquire_suggestion_db_updates_capability().await?;
```

Comment:
I wonder if not being able to get searcher DB updates should always prevent the whole project from being set up.
But perhaps it's good enough for now.

Reply:
We should design the exact behavior when we start thinking on handling lost connection on IDE start.